src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Split.cs

```diff
@@ -0,0 +1,136 @@
+        public static SpanSplitEnumerator<T> SplitAny<T>(this ReadOnlySpan<T> source, SearchValues<T> separators)
+            where T : IEquatable<T> => new SpanSplitEnumerator<T>(source, separators);
+
+        public ref struct SpanSplitEnumerator<T> where T : IEquatable<T>
+        {
+            private enum SplitMode
+            {
+                None = 0,
+                SingleToken,
+                Sequence,
+                Any,
+                SearchValues
+            }
+
+            private readonly ReadOnlySpan<T> _span;
+
+            private readonly T _separator = default!;
+            private readonly ReadOnlySpan<T> _separatorBuffer;
+            private readonly SearchValues<T> _searchValues = default!;
+
+            private readonly int _separatorLength;
+            private readonly SplitMode _splitMode;
+
+            private readonly bool _isInitialized = true;
+
+            private int _startCurrent = 0;
+            private int _endCurrent = 0;
+            private int _startNext = 0;
+
+            public SpanSplitEnumerator<T> GetEnumerator() => this;
+
+            public Range Current => new Range(_startCurrent, _endCurrent);
+
+            internal SpanSplitEnumerator(ReadOnlySpan<T> span, SearchValues<T> searchValues)
+            {
+                _span = span;
+                _separatorLength = 1;
+                _splitMode = SplitMode.SearchValues;
+                _searchValues = searchValues;
+            }
+
+            internal SpanSplitEnumerator(ReadOnlySpan<T> span, ReadOnlySpan<T> separator, bool treatAsSingleSeparator)
+            {
+                _span = span;
+                _separatorBuffer = separator;
+                _separatorLength = (_separatorBuffer.Length, treatAsSingleSeparator) switch
+                {
+                    (0, true) or (_, false) => 1,
+                    _ => separator.Length
+                };
+                _splitMode = treatAsSingleSeparator ? SplitMode.Sequence : SplitMode.Any;
+            }
+
+            internal SpanSplitEnumerator(ReadOnlySpan<T> span, T separator)
+            {
+                _span = span;
+                _separator = separator;
+                _separatorLength = 1;
+                _splitMode = SplitMode.SingleToken;
+            }
+
+            public bool MoveNext()
+            {
+                if (!_isInitialized || _startNext > _span.Length)
+                {
+                    return false;
+                }
+
+                ReadOnlySpan<T> slice = _span[_startNext..];
+
+                int separatorIndex = _splitMode switch
+                {
+                    SplitMode.SingleToken => slice.IndexOf(_separator),
+                    SplitMode.Sequence => slice.IndexOf(_separatorBuffer),
+                    SplitMode.Any => slice.IndexOfAny(_separatorBuffer),
+                    SplitMode.SearchValues => _searchValues.IndexOfAny(_span),
+                    _ => throw new UnreachableException()
```

Comment:
Should this throw or just silently fail?

Reply:
How would you get here such that it would matter? Assuming it's truly unreachable, I'd instead have an assert on the split mode before the switch and then just use _ for one of the valid cases.